eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java

```diff
@@ -104,39 +166,50 @@ public MessageBuilder() {
     public MessageBuilder(Optional<Integer> maxSize) {
       this.maxSize = maxSize;
 
-      size = MessageHeaderEncoder.ENCODED_LENGTH + MultiMessageEncoder.MessagesEncoder.HEADER_SIZE;
-
+      size = HEADER_SIZE;
     }
 
     public int getSize() {
       return size;
     }
 
-    public boolean addMessage(EventuateKafkaMultiMessageKeyValue message) {
-      int estimatedSize = estimatedMessageSize(message);
+    public boolean setHeaders(List<EventuateKafkaMultiMessagesHeader> headers) {
+      int estimatedSize = headers.stream().map(KeyValue::estimateSize).reduce(0, (a, b) -> a + b);
 
-      if (maxSize.map(ms -> size + estimatedSize > ms).orElse(false)) {
+      if (isSizeOverLimit(estimatedSize)) {
         return false;
       }
 
-      messagesToWrite.add(message);
+      this.headers = headers;
 
       size += estimatedSize;
 
       return true;
     }
 
-    private int estimatedMessageSize(EventuateKafkaMultiMessageKeyValue message) {
-      int keyLength = estimatedStringSizeInBytes(message.getKey());
-      int valueLength = estimatedStringSizeInBytes(message.getValue());
-      return 2 * 4 + keyLength + valueLength;
+    public boolean addMessage(EventuateKafkaMultiMessage message) {
+      int estimatedSize = message.estimateSize();
+
+      if (isSizeOverLimit(estimatedSize)) {
+        return false;
+      }
+
+      messagesToWrite.add(message);
+
+      size += estimatedSize;
+
+      return true;
     }
 
-    private int estimatedStringSizeInBytes(String s) {
-      return s == null ? 0 : s.length() * 2;
+    private boolean isSizeOverLimit(int estimatedSize) {
+      return maxSize.map(ms -> size + estimatedSize > ms).orElse(false);
     }
 
     public byte[] toBinaryArray() {
+      return toBinaryArray(false);
+    }
+
+    public byte[] toBinaryArray(boolean bufferAsIs) {
```

Comment:
What does `bufferAsIs` mean?

Reply:
I removed it by next commit.
But forgot to push.
Fixed it.
Used it to make result array length == encoded message size (not estimated)